# *PCA* - Comprimir imágenes

Una imagen digital estandar se obtiene apilando matrices de píxeles rojos, azules y verdes de intensidades comprendidas entre 0 y 255.

<img src="img/RGB.png">

Una imagen en escala de grises no contiene color, sino sólo tonos de gris. La intensidad de los píxeles de una imagen en escala de grises varía del negro (intensidad 0) al blanco (intensidad total 255), lo que la convierte en lo que solemos llamar una imagen en blanco y negro.

Digits dataset es un conjunto de datos de imágenes en escala de grises de dígitos manuscritos que contiene 1.797 imágenes de 8×8 píxeles.

In [1]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.datasets import load_digits

import warnings
warnings.filterwarnings('ignore')
 
digits = load_digits()
data = digits.data
data.shape

(1797, 64)

In [ ]:
data[0]

El módulo `sklearn.datasets` agiliza la importación de datos de dígitos importando de él la clase `load_digits`. La forma de los datos de dígitos es (1797, 64). Los píxeles de 8×8 se aplanan para crear un vector de longitud 64 para cada imagen.

In [ ]:
image_sample = data[100,:].reshape(8,8)
image_sample

In [ ]:
plt.imshow(image_sample, cmap='binary')

Ahora, utilizando *PCA*, reduzcamos las dimensiones de la imagen de 64 a sólo 2 para poder visualizar el conjunto de datos utilizando un gráfico de dispersión.

In [ ]:
from sklearn.decomposition import PCA

pca = PCA(2)
converted_data = pca.fit_transform(digits.data)

display(
    pca.explained_variance_ratio_,
    pca.explained_variance_ratio_.cumsum(), 
    converted_data.shape
    )

También podemos pasar un valor flotante menor que 1 en lugar de un número entero. Por ejemplo, `PCA(0.90)` significa que el algoritmo encontrará los componentes principales que explican el 90% de la varianza de los datos.

In [ ]:
plt.style.use('seaborn-whitegrid')
plt.figure(figsize = (10,6))
c_map = plt.cm.get_cmap('jet', 10)
plt.scatter(converted_data[:, 0], converted_data[:, 1], s = 15,
            cmap = c_map , c = digits.target)
plt.colorbar()
plt.xlabel('PC-1') , plt.ylabel('PC-2')
plt.show();

Una aplicación interesante del PCA es la compresión de imágenes. Echemos un vistazo a cómo podemos lograr esto con python.

In [ ]:
import cv2  # !pip install opencv-python
import numpy as np
from sklearn.decomposition import PCA

In [ ]:
img = cv2.imread('img/my_doggo_sample.jpeg')
img

In [ ]:
img

In [ ]:
plt.imshow(img);

In [ ]:
rgb_image = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
plt.imshow(rgb_image);

In [ ]:
img.shape

In [ ]:
img.size

In [ ]:
793*1280

In [ ]:
img

In [ ]:
# Separamos la imagen en los arrays R, G, B.
blue, green, red = cv2.split(img)

In [ ]:
blue.shape

In [ ]:
img_not_compressed = cv2.merge([red,green,blue])
plt.imshow(img_not_compressed);

In [ ]:
img_not_compressed = cv2.merge([blue, green, red])
plt.imshow(img_not_compressed);

OpenCV dividirá en canales Azul, Verde y Rojo en lugar de Rojo, Azul y Verde. Tenga mucho cuidado con la secuencia aquí.

In [ ]:
# Inicializamos el PCA con las 20 primeras componentes principales.
pca = PCA(20)
 
# Aplicando al canal rojo y luego aplicando la transformada inversa a la matriz transformada.
red_transformed = pca.fit_transform(red)
print('red_transformed.shape:', red_transformed.shape)
red_inverted = pca.inverse_transform(red_transformed)
print('red_inverted.shape:', red_inverted.shape)

# Aplicando al canal verde y luego aplicando la transformada inversa a la matriz transformada.
green_transformed = pca.fit_transform(green)
print('green_transformed.shape:', green_transformed.shape)
green_inverted = pca.inverse_transform(green_transformed)
print('green_inverted.shape:', green_inverted.shape)
 
# Aplicando al canal azul y luego aplicando la transformada inversa a la matriz transformada.
blue_transformed = pca.fit_transform(blue)
print('blue_transformed.shape:', blue_transformed.shape)
blue_inverted = pca.inverse_transform(blue_transformed)
print('blue_inverted.shape:', blue_inverted.shape)

En el proceso de reconstrucción de las dimensiones originales a partir de las dimensiones reducidas, se pierde algo de información, ya que sólo conservamos los componentes principales seleccionados, 20 en este caso.

In [ ]:
img_compressed = (cv2.merge([red_inverted, green_inverted, blue_inverted])).astype(np.uint8)

Apilando las matrices invertidas usando la función `dstack`. Aquí es importante especificar el tipo de datos de nuestras matrices, ya que la mayoría de las imágenes son de 8 bits. Cada píxel está representado por un byte de 8 bits.

In [ ]:
plt.imshow(img_compressed);

El resultado anterior es el que obtenemos cuando consideramos sólo 20 componentes principales.

Si aumentamos el número de componentes principales, la imagen de salida será más clara.

1. Ahora comprueba con cuántos componentes principales tus ojos no pueden ver la diferencia con el original.

1. El perro no debería ser tan azul, ¡arréglalo!

In [ ]:
# Inicializamos el PCA con las 300 primeras componentes principales.
pca = PCA(300)
 
# Aplicando al canal rojo y luego aplicando la transformada inversa a la matriz transformada.
red_transformed = pca.fit_transform(red)
print('red_transformed.shape:', red_transformed.shape)
red_inverted = pca.inverse_transform(red_transformed)
print('red_inverted.shape:', red_inverted.shape)

# Aplicando al canal verde y luego aplicando la transformada inversa a la matriz transformada.
green_transformed = pca.fit_transform(green)
print('green_transformed.shape:', green_transformed.shape)
green_inverted = pca.inverse_transform(green_transformed)
print('green_inverted.shape:', green_inverted.shape)
 
# Aplicando al canal azul y luego aplicando la transformada inversa a la matriz transformada.
blue_transformed = pca.fit_transform(blue)
print('blue_transformed.shape:', blue_transformed.shape)
blue_inverted = pca.inverse_transform(blue_transformed)
print('blue_inverted.shape:', blue_inverted.shape)

In [ ]:
img_compressed = (cv2.merge([red_transformed, green_transformed, blue_transformed])).astype(np.uint8)

#observamos la imagen comprimida
plt.imshow(img_compressed);

In [ ]:
pca.explained_variance_ratio_.cumsum()[99]

In [ ]:
pca.explained_variance_ratio_.cumsum()[299]

In [ ]:
# Inicializamos el PCA con las 280 primeras componentes principales.
pca = PCA(280)
 
# Aplicando al canal rojo y luego aplicando la transformada inversa a la matriz transformada.
red_transformed = pca.fit_transform(red)
print('red_transformed.shape:', red_transformed.shape)
red_inverted = pca.inverse_transform(red_transformed)
print('red_inverted.shape:', red_inverted.shape)

# Aplicando al canal verde y luego aplicando la transformada inversa a la matriz transformada.
green_transformed = pca.fit_transform(green)
print('green_transformed.shape:', green_transformed.shape)
green_inverted = pca.inverse_transform(green_transformed)
print('green_inverted.shape:', green_inverted.shape)
 
# Aplicando al canal azul y luego aplicando la transformada inversa a la matriz transformada.
blue_transformed = pca.fit_transform(blue)
print('blue_transformed.shape:', blue_transformed.shape)
blue_inverted = pca.inverse_transform(blue_transformed)
print('blue_inverted.shape:', blue_inverted.shape)